### PR TITLE
feat: Adding ToggleSwitch Material 3 Styles

### PR DESCRIPTION
**Pull request by eriklimakc**
﻿GitHub Issue: #806, closes https://github.com/unoplatform/uno/issues/9065

## PR Type

What kind of change does this PR introduce?
- Feature

## PR Checklist 
Please check if your PR fulfills the following requirements:

- Commits must be following the [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary)
- [X] Tested UWP
- [x] Tested iOS
- [X] Tested Android
- [X] Tested WASM
- [ ] Tested MacOS
- [X] Contains **No** breaking changes
  > If the pull request contains breaking changes, commit message must contain a detailed description of the action to take for the consumer of this library. As explained by the [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary)

### CODE REVIEW

**Comment by eriklimakc** [comment]
Currently getting a strange behavior on Skia GTK, may be related to https://github.com/unoplatform/uno/issues/8636


https://github.com/unoplatform/Uno.Themes/assets/114886335/bb787170-3e0e-486d-bce9-61fea5ab9550

**Comment by agneszitte** [comment]
> Currently getting a strange behavior on Skia GTK, may be related to [unoplatform/uno#8636](https://github.com/unoplatform/uno/issues/8636)
> 
>  gtk.m3.toggleswitch.mp4

@eriklimakc / @kazo0 Outside of GTK, are they additional issues with the current changes on the other platforms?
On Wasm I can see this issue when loading the page :
image

**Comment by eriklimakc** [comment]
> > Currently getting a strange behavior on Skia GTK, may be related to [unoplatform/uno#8636](https://github.com/unoplatform/uno/issues/8636)
> > gtk.m3.toggleswitch.mp4
> 
> @eriklimakc / @kazo0 Outside of GTK, are they additional issues with the current changes on the other platforms? On Wasm I can see this issue for example when loading the page : image

Yes, I'm currently working to fix this one. When `IsOn` is `true` the thumb starts in the wrong position for WASM/Android/GTK.

**Comment by eriklimakc** [comment]
Need this fix to proceed https://github.com/unoplatform/uno/issues/12643

**Comment by kazo0** [comment]
@Xiaoy312 is this at all related to https://github.com/unoplatform/uno/pull/12054?

**Review by agneszitte — CHANGES_REQUESTED** [review]
@eriklimakc Seems that there are other issues like these:
- **Visual State circle is off**
  - With current changes on WASM
    image
  - On The Figma file
    image
    
- **Make sure the icon is following the proper design rules for the size**
  - With the current changes on WASM
    image
  - On the Figma file
    image

- Make sure to use the proper brushes with different opacities for the different visual states
  Example in Figma here for the Visual State circle:
  image

(cc @NVLudwig, @kazo0)

**Comment by NVLudwig** [comment]
Very happy to see the icons are in there :)
Good comments from @agneszitte I would only add:
1) The animation seems a little fast and "flickery" compared to this https://m3.material.io/components/switch/guidelines#2495f276-189c-4d51-a9df-88e89e8e8185

https://github.com/unoplatform/Uno.Themes/assets/29584087/a47d3631-f68a-4334-a3b9-e4e5cc29d091


2) The spacing between the "thumb" and rail seems off, especially in the disabled states
**Published**
image
**Expected**
image

**Comment by eriklimakc** [comment]
# Dark and Light theme issues
It seems we have a problem with brushes when the Theme (Dark/Light) is changed. The colors are not matching.

## OS Theme as LIGHT

When OS theme is Light, it is correct: 
Light on System Light

If theme is changed to **Dark** on the Sample App we have:
Dark on System Light


## OS Theme as DARK

But, if the sample is launched with OS Theme as Dark we have:
Dark on System Dark


And when theme is changed on the Sample App to **Light**
Light on System Dark

Does this also happen with other controls?

@agneszitte @kazo0 @Xiaoy312

**Comment by kazo0** [comment]
@eriklimakc I believe this is fixed in a PR or is at least a known issue with the sample app, let me check

**Comment by NVLudwig** [comment]
Toggle Switch not showing at all for Material M3 on Safari

**Comment by eriklimakc** [comment]
> Toggle Switch not showing at all for Material M3 on Safari

@NVLudwig Did you get an empty screen instead or just the ToggleSwitch control not being displayed?

I tried it on my side on Safari for mac and Safari for iOS and it shows to me:

toggleswitch_safari_ios
toggleswitch_safari_mac

cc @agneszitte @kazo0

**Comment by NVLudwig** [comment]
@eriklimakc 
I just does not load the toggle switch for Material Tab it does in the Cupertino.

https://github.com/unoplatform/Uno.Themes/assets/29584087/cab17fe8-216e-4a4a-a442-e3edb5ed04a3

**Comment by eriklimakc** [comment]
# Current Material 3 ToggleSwitch:

## Android
android

## GTK
gtk

## iOS
ios

## UWP
uwp

## Wasm
wasm

**Comment by agneszitte** [comment]
@eriklimakc I am still seeing position issues for WASM for example with the latest changes
image
(cc @kazo0)

**Comment by eriklimakc** [comment]
> @eriklimakc I am still seeing position issues for WASM for example with the latest changes image (cc @kazo0)

For the disabled ones it's related to this issue https://github.com/unoplatform/uno/issues/12643.

cc @kazo0 @Xiaoy312

**Comment by NVLudwig** [comment]
Very good! @eriklimakc (works on safari in both tabs)

**Comment by eriklimakc** [comment]
@NVLudwig Thanks, I'm currently trying to fix the thumb shadow that's being clipped in some platforms, then the team can review all the code.

cc @agneszitte @kazo0

**Comment by NVLudwig** [comment]
Clipping issue with the inactive state Thumb
image

**Comment by iurycarlos** [comment]
Strange, for me it seems correct:

image
image

**Comment by NVLudwig** [comment]
I'm still getting strange clipping and it's spcifically happening when using a mouse then clicking and not moving the cursor away from the Control.

When you do this the thumb element goes to "Pressed" state for some reason at the end of the animation (as if you had held your finger down the whole time) and the "thumb" gets clipped every time. It resolves itself when you move the cursor away and the state goes back to "enabled" so you can't repro this in touch.

Here are some exemples

image
image
image

Happening on MacOS: 12.3.1 (21E258)
Safari: Version 15.4 (17613.1.17.1.13)

**Comment by agneszitte** [comment]
> I'm still getting strange clipping and it's spcifically happening when using a mouse then clicking and not moving the cursor away from the Control.
> 
> When you do this the thumb element goes to "Pressed" state for some reason at the end of the animation (as if you had held your finger down the whole time) and the "thumb" gets clipped every time. It resolves itself when you move the cursor away and the state goes back to "enabled" so you can't repro this in touch.
> 
> Here are some exemples
> 
> image image image
> 
> Happening on MacOS: 12.3.1 (21E258) Safari: Version 15.4 (17613.1.17.1.13)

Yeah I am not sure what is happening for Safari on mac @NVLudwig as these clipping issues are not reproducible on Chrome Mac.
@jeromelaban are you aware of similar clipping issues for Safari only ?

**Comment by agneszitte** [comment]
> Yeah I am not sure what is happening for Safari on mac @NVLudwig as these clipping issues are not reproducible on Chrome Mac. @jeromelaban are you aware of similar clipping issues for Safari only ?

Decided with @kazo0 that with @eriklimakc we will create a separate issue for Uno for the clipping issues related to Safari only.
@jeromelaban maybe think it is related to 4K but not sure.

**Review by kazo0 — APPROVED** [review]

**Comment by agneszitte** [comment]
> Azure Static Web Apps: Your stage site is ready! Visit it here: https://salmon-rock-0cfebe70f-1031.eastus2.azurestaticapps.net

@NVLudwig, can you please re-test Safari Mac with this link that includes the latest changes?
It is fixed normally for the clipping issues even on Safari, the only thing I noticed is that the animation is not as smooth as on Chrome Mac for Safari.

**Comment by NVLudwig** [comment]
Commented in Teams

File: src/library/Uno.Material/Styles/Controls/v2/ToggleSwitch.xaml [hunk 1]
```diff
@@ -2,662 +2,838 @@
 					xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
```
Comment (Xiaoy312) [inline src/library/Uno.Material/Styles/Controls/v2/ToggleSwitch.xaml:2]: you can trim some of the unused xmlnses here, especially the new ones added in this pr

File: src/library/Uno.Material/Styles/Controls/v2/ToggleSwitch.xaml [hunk 2]
```diff
@@ -2,662 +2,838 @@
...
+
+	<x:Double x:Key="KnobIconSize">16</x:Double>
+	<Thickness x:Key="KnobIconPadding">2</Thickness>
+	<Thickness x:Key="KnobOnMargin">2,0,0,0</Thickness>
+
+	<x:Double x:Key="SwitchKnobShadowSize">40</x:Double>
+	<Thickness x:Key="SwitchKnobOnShadowMargin">20,0,0,0</Thickness>
+	<Thickness x:Key="SwitchKnobOffShadowMargin">0,0,20,0</Thickness>
+
+	<!--Animations-->
```
Comment (Xiaoy312) [inline src/library/Uno.Material/Styles/Controls/v2/ToggleSwitch.xaml:40]: mvoe to AnimationConstants.xaml

**Comment by agneszitte** [comment]
> Commented in Teams

Will open a separate issue for Safari, thanks @NVLudwig

File: src/library/Uno.Material/Styles/Controls/v2/ToggleSwitch.xaml [hunk 3]
```diff
@@ -2,662 +2,838 @@
...
+	<x:String x:Key="ControlFastOutSlowInKeySpline">0.5,0,0.5,1</x:String>
+	<x:String x:Key="ControlNormalAnimationDuration">00:00:00.225</x:String>
+	<x:String x:Key="ControlFasterAnimationDuration">00:00:00.180</x:String>
+
+	<!--Converters-->
```
Comment (Xiaoy312) [inline src/library/Uno.Material/Styles/Controls/v2/ToggleSwitch.xaml:45]: can you prefix the xkey with the control name?

**Review by agneszitte — APPROVED** [review]

**Comment by agneszitte** [comment]
@mergifyio backport legacy/2x

**Pull request merged**